src/Microsoft.AspNetCore.Mvc.Core/Routing/ConsumesMatcherPolicy.cs

```diff
@@ -0,0 +1,242 @@
+            for (var i = 0; i < endpoints.Count; i++)
+            {
+                var endpoint = endpoints[i];
+                var contentTypes = endpoint.Metadata.GetMetadata<IConsumesMetadata>()?.ContentTypes ?? Array.Empty<string>();
+                if (contentTypes.Count == 0)
+                {
+                    // OK this means that this endpoint matches *all* content methods.
+                    // So, loop and add it to all states.
+                    foreach (var kvp in edges)
+                    {
+                        kvp.Value.Add(endpoint);
+                    }
+                }
+                else
+                {
+                    // OK this endpoint matches specific content types -- we have to loop through edges here
+                    // because content types could either be exact (like 'application/json') or they
+                    // could have wildcards (like 'text/*'). We don't expect wildcards to be especially common
+                    // with consumes, but we need to support it.
+                    for (var j = 0; j < contentTypes.Count; j++)
+                    {
+                        var contentType = contentTypes[j];
+
+                        var mediaType = new MediaType(contentType);
+
+                        foreach (var kvp in edges)
+                        {
+                            // The edgeKey maps to a possible request header value
+                            var edgeKey = new MediaType(kvp.Key);
+
+                            // Example: 'application/json' is subset of 'application/*'
+                            // 
+                            // This means that when the request has content-type 'application/json' an endpoint
+                            // what consumes 'application/*' should match.
+                            if (edgeKey.IsSubsetOf(mediaType))
+                            {
+                                kvp.Value.Add(endpoint);
+                            }
+                        }
+                    }
+                }
+            }
+
+            // If after we're done there isn't any endpoint that accepts */*, then we'll synthesize an
+            // endpoint that always returns a 405.
+            if (!edges.ContainsKey(AnyContentType))
+            {
+                edges.Add(AnyContentType, new List<Endpoint>()
+                {
+                    CreateRejectionEndpoint(),
+                });
+            }
+
+            return edges
+                .Select(kvp => new PolicyNodeEdge(kvp.Key, kvp.Value))
+                .ToArray();
+        }
+
+        private Endpoint CreateRejectionEndpoint()
+        {
+            return new MatcherEndpoint(
+                (next) => (context) =>
+                {
+                    context.Response.StatusCode = StatusCodes.Status415UnsupportedMediaType;
+                    return Task.CompletedTask;
+                },
+                RoutePatternFactory.Parse("/"),
```

Comment:
What do you think about adding some kind of `RoutePatternFactory.EmptyPattern` property for situations like this? Or `RoutePattern.Empty`.

Alternatively could add it to this class to reuse.

Reply:
I'm not a huge fan of this idea, I think it will appear anywhere outside of test code. 

Also this won't be `MatcherEndpoint` for long https://github.com/aspnet/Routing/issues/648